### DISCUSSION TITLE
Is it possible to create a nitro plugin inside a Nuxt Module?

### DISCUSSION THREAD

**Discussion by leifmarcus** [nuxt/nuxt#16688] [Questions] [answered]
Currently, when creating a Nitro Plugin inside a module, the `defineNitroPlugin` is undefined.

```ts
// entry of the nuxt module:
// -------------------------
export default defineNuxtModule({
  // ...
  setup() {
    // inside the module a plugin is added which works as expected
    const { resolve } = createResolver(import.meta.url);
    addPlugin({ src: resolve(runtimeDir, 'NitroPlugin.ts'), mode: 'server' });
  }
});

// A simple plugin file looks like:
// --------------------------------
export default defineNitroPlugin((nitroApp) => {
  nitroApp.hooks.hook('render:html', (html) => {
    html.head.push('some custom html to append');
  });
});
```
In this case the `defineNitroPlugin` is `undefined`. It seems like it cannot be resolved correctly. But if I'm importing it from `import { defineNitroPlugin } from 'nitropack/dist/runtime/plugin';` as described in [this comment](https://github.com/nuxt/framework/issues/5565#issuecomment-1239422984), the plugin is not working and I get the following warning:

>You are using a plugin that has not been wrapped in `defineNuxtPlugin`. It is advised to wrap your plugins as in the future this may enable enhancements

My Question is, if it is possible to create a Nitro plugin inside a Nuxt Module or if there is another possibility to add an html string into the Head of the page from inside a Nuxt Module. The reason for that is, that in the project I'm working on, the design system provides some html markup for the header of the page. This functionality should be shared with multiple teams. This is why I decided to put this into a shared nuxt module. But as it seems, there is no possibility to do so in nuxt.

**Comment by TheAlexLichter** [comment, marked as answer]
Check out https://github.com/nuxt/nuxt/discussions/16780 regarding adding a Nitro plugin via Nuxt module.
A Nitro plugin (e.g. shown in https://github.com/nuxt/nuxt/discussions/16743#discussioncomment-3741384) should be the best option if it is just markup that does not live in components 👍🏻

**Reply by leifmarcus** [reply]
@manniL Thanks a lot for the answer. I'll have a look into the issues you linked.

**Reply by leifmarcus** [reply]
@manniL For me the implementation that is mentioned in https://github.com/nuxt/nuxt/discussions/16743#discussioncomment-3741384 works as expected!

**Comment by bitbytebit1** [comment]
Oops, can we get a reference here of the solution? The links are dead

**Reply by TheAlexLichter** [reply]
Links updated 👍🏻

**Reply by bitbytebit1** [reply]
Thank you!

**Comment by YannickArmspach** [comment]
Simply use "addServerPlugin" function in your module.ts file (https://nuxt.com/docs/api/kit/nitro#addserverplugin)
